arch/cortex-m0/src/lib.rs

```diff
@@ -31,8 +31,12 @@ struct HardFaultStackedRegisters {
 }
 
 #[cfg(all(target_arch = "arm", target_os = "none"))]
-#[inline(never)]
-unsafe fn kernel_hardfault(faulting_stack: *mut u32) {
+/// Handle a hard fault that occurred in the kernel. This function is invoked
+/// by the naked hard_fault_handler function.
+///
+/// TODO: It should ensure that in case of a stack overflow, this function
+/// executes on a clean stack. Is that the case right now?
```

Comment:
I don't see how the Cortex-M0 hard fault handler makes enough stack space to handle a fault in a stack-overflow case. Can someone enlighten me how this works, or whether this is indeed missing right now?

Reply:
IIRC, The `panic!` macro resets the stack [same trick is used in v7m, go straight to panic on the stack overflow case].

On quick glance, the relies on the `hardfault_stacked_regsiters` not being put on the local stack — that... may or may not actually work [in the v7m case, it's a static string passed to panic, so presumably it's just loading a pointed and jumping].

Maybe update the TODO here to explain the assumption, but I don't think this needs to be fixed with this PR.

Reply:
> IIRC, The `panic!` macro resets the stack [same trick is used in v7m, go straight to panic on the stack overflow case].

From a brief look at the generated assembly I can neither confirm or deny this. I only brought this up given that on ARMv7m we actually have dedicated assembly to reset the stack before calling into `panic!` (or I'm reading things very wrong):

https://github.com/tock/tock/blob/0466e1ef56822c300c3fe81e706443fd1baee7f3/arch/cortex-m/src/lib.rs#L727-L732

I was missing that same assembly on v6m.

> On quick glance, this relies on the `hardfault_stacked_regsiters` not being put on the local stack — that... may or may not actually work [in the v7m case, it's a static string passed to panic, so presumably it's just loading a pointer and jumping].

Right, that seems tricky. If `panic!` were to reset the stack, we could be lucky that this debug information is sufficiently far down the stack that it doesn't collide with the panic itself. That seems ... brittle.

> Maybe update the TODO here to explain the assumption, but I don't think this needs to be fixed with this PR.

Right. In summary I still don't feel like I have a good grasp of what's going on, but I agree that this is a separate issue.